core/server/master/src/main/java/alluxio/master/metrics/MetricsStore.java

```diff
@@ -111,21 +107,55 @@ public void putWorkerMetrics(String hostname, List<Metric> metrics) {
    * @param clientId the id of the client
    * @param metrics the new metrics
    */
-  public void putClientMetrics(String hostname, String clientId,
-      List<Metric> metrics) {
+  public void putClientMetrics(String hostname, String clientId, List<Metric> metrics) {
     if (metrics.isEmpty()) {
       return;
     }
     LOG.debug("Removing metrics for id {} to replace with {}", clientId, metrics);
     synchronized (mClientMetrics) {
-      mClientMetrics.removeByField(ID_INDEX, getFullInstanceId(hostname, clientId));
-      for (Metric metric : metrics) {
-        if (metric.getHostname() == null) {
-          continue; // ignore metrics whose hostname is null
-        }
-        mClientMetrics.add(metric);
+      putReportedMetrics(mClientMetrics, getFullInstanceId(hostname, clientId), metrics);
+    }
+  }
+
+  /**
+   * Update the reported metrics with the given instanceId and set of metrics received from a
+   * worker or client.
+   *
+   * Any metrics from the given instanceId which are not reported in the new set of metrics are
```

Comment:
Why do we remove the old metrics if they are not contained in the new set?

Reply:
This was the previous behavior of the metrics heartbeat. The behavior itself may be questionable, but I didn't want to modify it for the scope of this PR.

One issue, I think, is that the metric hash code depends upon the value of the metric itself. If we're getting different values over heartbeats, its possible to store more than one metric of the same type if the previous isn't removed. I, however, don't know if there is ever a case where the lists of worker metrics differs between heartbeats (whether they become a subset/superset).